tasks/broad/ImputationTasks.wdl

```diff
@@ -0,0 +1,752 @@
+    preemptible : 3
+  }
+
+  output {
+    File aggregated_metrics = "~{basename}_aggregated_imputation_metrics.tsv"
+  }
+}
+
+task StoreChunksInfo {
+  input {
+    Array[String] chroms
+    Array[Int] starts
+    Array[Int] ends
+    Array[Int] vars_in_array
+    Array[Int] vars_in_panel
+    Array[Boolean] valids
+    String basename
+    String rtidyverse_docker
+  }
+
+  command <<<
+    Rscript -<< "EOF"
+    library(dplyr)
+    library(readr)
+
+    chunk_info <- tibble(chrom = c("~{sep='", "' chroms}"), start = c("~{sep='", "' starts}"), ends = c("~{sep='", "' ends}"), vars_in_array = c("~{sep='", "' vars_in_array}"), vars_in_panel = c("~{sep='", "' vars_in_panel}"), chunk_was_imputed = as.logical(c("~{sep='", "' valids}")))
+    failed_chunks <- chunk_info %>% filter(!chunk_was_imputed) %>% select(-chunk_was_imputed)
+    n_failed_chunks <- nrow(failed_chunks)
+    write_tsv(chunk_info, "~{basename}_chunk_info.tsv")
+    write_tsv(failed_chunks, "~{basename}_failed_chunks.tsv")
+    write(n_failed_chunks, "n_failed_chunks.txt")
+    EOF
+  >>>
+
+  runtime {
+    docker: rtidyverse_docker
+    preemptible : 3
+  }
+
+  output {
+    File chunks_info = "~{basename}_chunk_info.tsv"
+    File failed_chunks = "~{basename}_failed_chunks.tsv"
+    Int n_failed_chunks = read_int("n_failed_chunks.txt")
+  }
+}
+
+task MergeImputationQCMetrics {
+  input {
+    Array[File] metrics
+    String basename
+    String rtidyverse_docker
+  }
+
+  Int disk_size = 100 + ceil(size(metrics, "GiB"))
+
+  command <<<
+    Rscript -<< "EOF"
+    library(dplyr)
+    library(readr)
+    library(purrr)
+    library(ggplot2)
+
+    metrics <- list("~{sep='", "' metrics}") %>% map(read_tsv) %>% reduce(`+`) %>% mutate(frac_sites_r2_gt_0.3=total_sites_r2_gt_0.3/total_sites, frac_sites_with_var_r2_gt_0.3=total_sites_r2_gt_0.3/total_sites_with_var)
+
+    write_tsv(metrics, "~{basename}_aggregated_imputation_metrics.tsv")
+    write(metrics %>% pull(frac_sites_with_var_r2_gt_0.3), "frac_well_imputed.txt")
+
+    EOF
+  >>>
+
+  runtime {
+    docker: rtidyverse_docker
+    disks : "local-disk " + disk_size + " HDD"
+    preemptible : 3
+  }
+
+  output {
+    File aggregated_metrics = "~{basename}_aggregated_imputation_metrics.tsv"
+    Float frac_well_imputed = read_float("frac_well_imputed.txt")
+  }
+}
+
+task SetIDs {
+  input {
+    File vcf
+    File vcfIndex
+    String output_basename
+    String bcftools_docker
+  }
+
+  Int disk_size = 100 + ceil(2.2 * size(vcf, "GiB"))
+
+  command <<<
+    bcftools annotate ~{vcf} --set-id '%CHROM\:%POS\:%REF\:%FIRST_ALT' -Oz -o ~{output_basename}.vcf.gz
+    bcftools index -t ~{output_basename}.vcf.gz
+  >>>
+
+  runtime {
+    docker: bcftools_docker
+    disks: "local-disk " + disk_size + " HDD"
+    memory: "4 GiB"
+  }
+
+  output {
+    File output_vcf = "~{output_basename}.vcf.gz"
+    File output_vcf_index = "~{output_basename}.vcf.gz.tbi"
+  }
+}
+
+task ExtractIDs {
```

Comment:
Could some of these smaller tasks that are using bcftools be combined into a single task?

Reply:
I think I want to keep the ExtractIDs job separate from adding the IDs to the vcf, since having the ids in the vcf is useful for downstream tools.  But can definitely combine the SetIds and SortIds tasks.